libs/core/thread_support/include/hpx/thread_support/spinlock.hpp

```diff
@@ -45,10 +46,11 @@ namespace hpx { namespace util { namespace detail {
         void lock() noexcept
         {
             // Optimistically assume the lock is free on the first try
-            if (!m.exchange(true, std::memory_order_acquire))
+            if (try_lock())
```

Comment:
Is there now any reason why this spinlock should first do `try_lock` and the other one would first check `is_locked`? What's different?

Reply:
There's no difference. The first one does a `try_lock` in hopes of getting the lock. If the `load` operation returns false it continues executing without doing an exchange (that we are trying to reduce). The second one does exactly the same. It looks different because `yield_k` functions are different in both implementations. The first one takes a counter k and implements a backoff strategy. The second one takes a predicate and does what the first one does but only until the predicate returns true. So, the design for both of them is essentially the same. The function signature causes them to look different.

Reply:
Let me rephrase: Why not structure this exactly the same way as the spinlock in `synchronization`, with `is_locked` etc.?

Additionally, with the change to use `try_lock` in the `if` these two are identical:
```
if (try_lock) return;
unsigned k = 0;
do { yield_k(k++); } while (!try_lock());
```
and:
```
unsigned k = 0;
while (!try_lock) { yield_k(k++); }
```

Reply:
Aah, right! I got what you're saying now. Yes, it can be structured this way and I should do it do, apologies.

Reply:
Sorry @NK-Nikunj for beating this topic to death, but:

`synchronization/spinlock.hpp` was changed from
```
            while (!acquire_lock())
            {
                util::yield_while([this] { return is_locked(); },
                    "hpx::lcos::local::spinlock::lock");
            };
```
to
```
            do
            {
                util::yield_while([this] { return is_locked(); },
                    "hpx::lcos::local::spinlock::lock");
            } while (!acquire_lock());
```
because of `<reasons>`.

`thread_support/spinlock.hpp` is now:
```
            unsigned k = 0;
            while (!try_lock())
            {
                yield_k(k++);
            };
```

Now my original question in this thread becomes a bit more obvious: Are these two still equivalent, and if not, why should they not be?

In https://github.com/STEllAR-GROUP/hpx/pull/5445#discussion_r669503863 I merely wanted to point out that if there is going to be an `if (x) return; do { z; } while (!x);` it's better written as `while (!x) { z; };` not necessarily suggest that it should be a `while (!x) { z; };`

Reply:
> Are these two still equivalent, and if not, why should they not be?

Yes. I think the confusion arises from the behavior of yield functions. Let's consider both cases one by one. For `thread_support/spinlock.hpp`, we do:

```
unsigned k = 0;
while (!try_lock())
{
    yield_k(k++);
};
```

This does one atomic load which when computed to true does a yield of 0 following which it does another atomic load. If that computes to true as well then we do yield of 1 and so on. In essence, we are doing atomic loads as long as yield either suspends the thread or the atomic loads computes to false. Finally, if it computes to false, it does an atomic exchange to try and acquire the lock.

Now consider `synchronization/spinlock.hpp` file with `do {} while` implementation. Here, it goes into yield first (the internal yield variable k is still 0) and tries for an atomic load (defined by `is_locked()`). It continues to yield with k = 1, 2, ... until `is_locked` returns false following which it does an atomic exchange to acquire the lock (defined by `!acquire_lock()`). So, it has the same functioning as the above case.

Now consider the case with `while` instead of `do {} while`. Here we do an extra atomic load in the while loop. So, for k = 0, we do one atomic load when we enter the loop followed by another atomic load through `is_locked()`. Now, this wouldn't matter for scenarios where the first thread acquires the lock (as we never get into the while loop) but when the thread has to spin on while, we end up doing an extra load.

Again, the cost of a single atomic load might not be substantial, but I won't call them equivalent implementations then :P

Reply:
We should avoid unneeded atomic operations in any case. They might create non-measurable overheads, but they are still atomics (even if if they use relaxed semantics), i.e. they force the core's execution pipeline to flush/stall.

Reply:
For posterity, we decided to attempt the unification of the two spinlock implementations separately. This is resolved for now.